server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -564,234 +423,116 @@ public FlatObjectFieldType fieldType() {
-                  It reads the JSON object and parsed to a list of string
-                 */
-                XContentParser parser = jsonToStringParser.parseObject();
-
-                XContentParser.Token currentToken;
-                while ((currentToken = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
-                    switch (currentToken) {
-                        case FIELD_NAME:
-                            fieldName = parser.currentName();
-                            break;
-                        case VALUE_STRING:
-                            String value = parser.textOrNull();
-                            parseValueAddFields(context, value, fieldName);
-                            break;
-                    }
-
-                }
             }
+            parseObject(ctxParser, context);
         }
-
     }
 
     @Override
-    public Iterator<Mapper> iterator() {
-        List<Mapper> subIterators = new ArrayList<>();
-        if (valueFieldMapper != null) {
-            subIterators.add(valueFieldMapper);
-        }
-        if (valueAndPathFieldMapper != null) {
-            subIterators.add(valueAndPathFieldMapper);
-        }
-        if (subIterators.size() == 0) {
-            return super.iterator();
-        }
-        @SuppressWarnings("unchecked")
-        Iterator<Mapper> concat = Iterators.concat(super.iterator(), subIterators.iterator());
-        return concat;
+    protected String contentType() {
+        return CONTENT_TYPE;
     }
 
-    /**
-     * parseValueAddFields method will store data to Lucene.
-     * the JsonToStringXContentParser returns XContentParser with 3 string fields
-     * fieldName, fieldName._value, fieldName._valueAndPath.
-     * parseValueAddFields recognized string by the stringfield name,
-     * fieldName will be store through the parent FlatObjectFieldMapper,which contains all the keys
-     * fieldName._value will be store through the valueFieldMapper, which contains the values of the Json Object
-     * fieldName._valueAndPath will be store through the valueAndPathFieldMapper, which contains the "path=values" format
-     */
-    private void parseValueAddFields(ParseContext context, String value, String fieldName) throws IOException {
+    private void parseObject(XContentParser parser, ParseContext context) throws IOException {
+        assert parser.currentToken() == XContentParser.Token.START_OBJECT;
+        parser.nextToken(); // Skip the outer START_OBJECT. Need to return on END_OBJECT.
 
-        NamedAnalyzer normalizer = fieldType().normalizer();
-        if (normalizer != null) {
-            value = normalizeValue(normalizer, name(), value);
+        LinkedList<String> path = new LinkedList<>(Collections.singleton(fieldType().name()));
+        HashSet<String> leafPaths = new HashSet<>();
+        while (parser.currentToken() != XContentParser.Token.END_OBJECT) {
```

Comment:
In JsonToStringXContentParser, we can deduplication the fieldName,valueList,valueAndPathList.